### PR TITLE
Migrations

### PR DESCRIPTION
**Pull request by WouterSioen**
A migrations implementation for Fork CMS. Note that I have not a lot of ruby/rake experience, so this could probably be optimized a lot.

# how does it work?

When you have a directory structure like this

    Fork-project-root
      |- migrations
          |- migrations-folder-1
              |- locale.xml
              |- update.php
              |- update.sql 

The locale file will be imported, the sql will be executed on the database and the php script will be run. It works like this:

### deploy:setup

* executed_migrations file will be created
* maintenance page will be created on the server

### first deploy

* executed_migrations will be filled with the migration files that are already available

### other deploys

* we check if there are new migrations folders (next steps assume migrations folders where found)
* the maintenance page is linked
* a database backup is taken
* we loop trough migrations and execute them
* migrations are added to the executed_migrations file
* maintenance page is replaced by application again

### on rollback during execution of migrations

* database backup will be put back
* maintenance page is replaced by application again.

### CODE REVIEW

**Comment by WouterSioen** [comment]
Note that this pull request depends on https://github.com/sumocoders/forkcms/pull/33

**Comment by WouterSioen** [comment]
The maintenance page is Fork branded

screen shot 2015-02-19 at 15 51 09

**Comment by turanct** [comment]
ziet er strak uit

**Comment by tijsverkoyen** [comment]
Ziet er goed uit.
@tomkurt kan jij eens naar de ruby-code kijken en je mening geven

**Comment by WouterSioen** [comment]
poke @tomkurt

File: lib/forkcms_deploy/forkcms_3.rb [hunk 1]
```diff
@@ -99,6 +106,136 @@
...
+			migrationsFileExists = capture("if [ -f #{shared_path}/executed_migrations ]; then echo 'yes'; fi").chomp
+
+			# Only create the file if it doesn't exists.
+			unless migrationsFileExists == 'yes'
+				# Create an empty executed_migrations file
+				put '', "#{shared_path}/executed_migrations"
+			end
+
+			# Create a maintenance folder containing the index page from our gem
+			maintenance_path = File.dirname(__FILE__)
```
Comment (tomkurt) [inline lib/forkcms_deploy/forkcms_3.rb]: Wordt niet gebruikt

File: lib/forkcms_deploy/forkcms_3.rb [hunk 2]
```diff
@@ -99,6 +106,136 @@
...
+			maintenance_path = "#{maintenance_path}/../maintenance"
+			run "mkdir #{shared_path}/maintenance"
+
+			# copy the contents of the index.html file to our shared folder
+			indexfile = File.open("#{maintenance_path}/index.html", "rb")
+			put indexfile.read, "#{shared_path}/maintenance/index.html"
+			indexfile.close
```
Comment (tomkurt) [inline lib/forkcms_deploy/forkcms_3.rb]: kan mooier met block 
File.open('..') do |f|
  f.read
end

File: lib/forkcms_deploy/forkcms_3.rb (hunk 1 above)
Comment (WouterSioen) [inline lib/forkcms_deploy/forkcms_3.rb]: Jawel, op lijn 130 staat dit: `indexfile = File.open("#{maintenance_path}/index.html", "rb")`

**Comment by WouterSioen** [comment]
feedback handled

**Comment by tijsverkoyen** [comment]
Mag wat mij betreft gemerged worden

**Pull request merged**